File: app/src/main/java/com/illiad/troad/service/channel/FildesChannel.java
package com.illiad.troad.service.channel;

import io.netty.channel.*;

import java.io.File;
import java.io.FileDescriptor;
import java.io.IOException;
import java.net.SocketAddress;
import java.nio.channels.AsynchronousFileChannel;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;

public class FildesChannel extends AbstractChannel { // Extends AbstractChannel

    private static final ChannelMetadata METADATA = new ChannelMetadata(false); // Example metadata
    private final FildesChannelConfig config = new FildesChannelConfig(this); // Custom config
    private final File file;
    private AsynchronousFileChannel afc;

    public FildesChannel(Channel parent, File file) {
        super(parent);
        this.file = file;

    }

    @Override
    public ChannelMetadata metadata() {
        return METADATA;
    }

    @Override
    public ChannelConfig config() {
        return config;
    }

    @Override
    public boolean isOpen() {
        return afc != null && afc.isOpen();
    }

    @Override
    public boolean isActive() {
        return isOpen(); // Active when the file channel is open
    }

    @Override
    protected AbstractUnsafe newUnsafe() {
        return new FildesChannelUnsafe();
    }

    @Override
    protected boolean isCompatible(EventLoop loop) {
        // You might need to check if the EventLoop is compatible with your custom channel's operations
        return true; // For simplicity in this example
    }

    @Override
    protected SocketAddress localAddress0() {
        // Return a representation of the file path as a SocketAddress
        return new FileChannelAddress(file);
    }

    @Override
    protected SocketAddress remoteAddress0() {
        // File channels don't typically have a remote address
        return null;
    }

    @Override
    protected void doBind(SocketAddress localAddress) throws Exception {
        // Implement binding logic (if needed)
        // In this case, binding might involve associating the channel with the file path
    }

    @Override
    protected void doDisconnect() throws Exception {
        // This method is called when the Channel's disconnect() method is invoked.
        // Here, you should perform actions to "disconnect" your file channel,
        // which typically means releasing resources associated with the file.
        System.out.println("FildesChannel: doDisconnect() called.");
        if (afc != null) {
            try {
                afc.close(); // Close the file channel
                afc = null;
            } catch (IOException e) {
                // Handle potential IOException during close
                throw new ChannelException("Failed to close file channel", e);
            }
        }
        // After successfully disconnecting, Netty will typically trigger a channelInactive event.
    }

    // You also need to implement the doClose() method to handle channel closure.
    @Override
    protected void doClose() throws Exception {
        // This method is called when the Channel's close() method is invoked.
        // This method should also release resources and mark the channel as closed.
        System.out.println("CustomFileChannel: doClose() called.");
        if (afc != null) {
            try {
                afc.close();
                afc = null;
            } catch (IOException e) {
                throw new ChannelException("Failed to close file channel", e);
            }
        }
        // After doClose() completes, Netty will fire the channelInactive event
        // followed by channelUnregistered if not associated with an EventLoop.
    }

    @Override
    protected void doBeginRead() throws Exception {
        // Implement the logic to initiate reading from the file channel
        // This might involve scheduling a task on the EventLoop to read from the file
        // and fire channelRead events
    }

    @Override
    protected void doWrite(ChannelOutboundBuffer in) throws Exception {
        // Implement writing logic if needed
    }

    // Example of a custom SocketAddress for your file channel
    private static final class FileChannelAddress extends SocketAddress {
        private final File file;

        FileChannelAddress(File file) {
            this.file = file;
        }

        @Override
        public String toString() {
            return "file:" + file.getAbsolutePath();
        }
    }

    // Unsafe implementation to handle I/O operations
    private final class FildesChannelUnsafe extends AbstractUnsafe {

        @Override
        public void connect(SocketAddress remoteAddress, SocketAddress localAddress, ChannelPromise promise) {
            // Implement connection logic (if needed)
            // In this case, connecting might involve opening the AsynchronousFileChannel
            try {
                afc = AsynchronousFileChannel.open(file.toPath(), StandardOpenOption.READ, StandardOpenOption.WRITE);
                // Signal success
                promise.setSuccess();
            } catch (IOException e) {
                // Signal failure
                promise.setFailure(e);
            }
        }

        // Implement other Unsafe methods for file channel interactions
        // ...
    }

    private static Path fd2Path(FileDescriptor fd) {

        int fdNum = fd.;
        return Paths.get("/proc/self/fd/");
    }

    // Custom ChannelConfig for your file channel
    private static final class FildesChannelConfig extends DefaultChannelConfig {
        FildesChannelConfig(Channel channel) {
            super(channel, new AdaptiveRecvByteBufAllocator());
        }

        // Implement custom configuration options for your file channel
        // ...
    }

    // Custom SocketAddress to represent the file path
    private static final class FildesChannelAddress extends SocketAddress {
        private final File file;

        public FildesChannelAddress(File file) {
            this.file = file;
        }

        @Override
        public String toString() {
            return "FileChannelAddress(" + file.getAbsolutePath() + ")";
        }
    }
}

File: app/src/main/java/com/illiad/troad/service/channel/FildesChannel.kt
package com.illiad.troad.service.channel

import android.util.Log
import io.netty.buffer.ByteBuf
import io.netty.channel.*
import io.netty.util.internal.logging.InternalLoggerFactory
import java.io.FileDescriptor
import java.io.IOException
import java.net.SocketAddress
import java.nio.channels.FileChannel
import java.util.concurrent.TimeUnit

class FildesChannel(parent: Channel?, private val fd: FileDescriptor) : AbstractChannel(parent) {

    private val logger = InternalLoggerFactory.getInstance(FildesChannel::class.java)
    private val config = FildesChannelConfig(this)

    // Single bidirectional channel for the TUN FD
    private var nioChannel: FileChannel? = null

    @Volatile
    private var isPaused = false
    private val ZERO_READ_PAUSE_MS = 10L

    init {
        try {
            if (!fd.valid()) throw IOException("Invalid FileDescriptor")

            // Use FileInputStream so the NIO FileChannel marks itself as READABLE.
            // Because the TUN FD is bidirectional, this channel WILL still allow writes.
            this.nioChannel = java.io.FileInputStream(fd).channel

            Log.d("FildesChannel", "NIO Channel initialized from FileInputStream for R/W")
        } catch (e: Exception) {
            Log.e("FildesChannel", "Failed to initialize NIO channel from FD", e)
            close(voidPromise())
        }
    }

    override fun config(): FildesChannelConfig = config
    override fun metadata(): ChannelMetadata = ChannelMetadata(false)
    override fun isRegistered(): Boolean = super.isRegistered()
    override fun isActive(): Boolean = fd.valid() && nioChannel?.isOpen == true
    override fun isOpen(): Boolean = fd.valid()
    override fun isCompatible(loop: EventLoop?): Boolean {
        return loop is EventLoop
    }


    // --- Outbound Logic: Injecting packets back into TUN ---
    override fun doWrite(buf: ChannelOutboundBuffer) {
        val channel = nioChannel ?: return

        while (true) {
            val msg = buf.current() ?: break
            if (msg is ByteBuf) {
                try {
                    val localBuf = msg.nioBuffer()
                    val written = channel.write(localBuf)
                    if (written <= 0) break // TUN is temporarily full
                    buf.remove() // Successfully injected packet
                } catch (e: Exception) {
                    buf.remove(e)
                }
            } else {
                buf.remove(UnsupportedOperationException("Only ByteBuf is supported"))
            }
        }
    }

    override fun doBeginRead() {
        // This is called by the pipeline or by Unsafe.beginRead()
        // We delegate to our internal unsafe to perform the actual I/O
        (unsafe() as FildesUnsafe).readFromFildes()
    }

    // --- Inbound Logic: Captured through this custom Unsafe ---
    override fun newUnsafe(): AbstractUnsafe = FildesUnsafe()

    private inner class FildesUnsafe : AbstractUnsafe() {

        // Rename this from 'beginRead' to something unique like 'readFromFildes'
        fun readFromFildes() {
            if (isPaused || !isActive) return

            val allocHandle = recvBufAllocHandle()
            allocHandle.reset(config())
            val allocator = config().allocator
            var firedRead = false

            try {
                do {
                    val byteBuf = allocHandle.allocate(allocator)
                    // Use nioChannel directly
                    val bytesRead = byteBuf.writeBytes(nioChannel!!, byteBuf.writableBytes())
                    allocHandle.lastBytesRead(bytesRead)

                    if (bytesRead <= 0) {
                        byteBuf.release()
                        if (bytesRead < 0) {
                            close(voidPromise()) // EOF
                        } else {
                            handleZeroRead() // 0 bytes available
                        }
                        break
                    }

                    firedRead = true
                    allocHandle.incMessagesRead(1)
                    pipeline().fireChannelRead(byteBuf)

                } while (allocHandle.continueReading())

            } catch (t: Throwable) {
                pipeline().fireExceptionCaught(t)
            } finally {
                allocHandle.readComplete()
                if (firedRead) {
                    pipeline().fireChannelReadComplete()
                }
            }
        }

        private fun handleZeroRead() {
            isPaused = true
            eventLoop().schedule({
                isPaused = false
                if (config().isAutoRead) {
                    // Call the top-level doBeginRead again
                    this@FildesChannel.doBeginRead()
                }
            }, ZERO_READ_PAUSE_MS, TimeUnit.MILLISECONDS)
        }

        // Required override for AbstractUnsafe
        override fun connect(
            remote: SocketAddress?,
            local: SocketAddress?,
            promise: ChannelPromise?
        ) {
            promise?.setFailure(UnsupportedOperationException("TUN is point-to-point"))
        }
    }

    override fun doClose() {
        nioChannel?.close()
    }

    override fun doBind(localAddress: SocketAddress?) {}
    override fun doDisconnect() {
        doClose()
    }

    override fun localAddress0(): SocketAddress? = null
    override fun remoteAddress0(): SocketAddress? = null
}


